packages/rest/src/rest.server.ts

```diff
@@ -717,6 +742,10 @@ export class RestServer extends Context implements Server, HttpServerLike {
       spec = this.updateSpecFromRequest(spec, requestContext);
     }
 
+    // Apply OAS enhancers to the OpenAPI specification
+    this.OASEnhancer.spec = spec;
+    spec = await this.OASEnhancer.applyEnhancerByName('security');
```

Comment:
I don't think the REST server should be invoking specific enhancers defined by a name, because that makes it impossible for other extensions to contribute new enhancers.

As I understand the proposed design:
1. The REST API invokes all registered API enhancers via `applyAllEnhancers()`.
2. Extensions like `@loopback/authentication` contribute custom API enhancers to provide additional OpenAPI metadata like the authentication scheme.

The REST component (`@loopback/rest`) should not include any security schemas in the OpenAPI spec out of the box, simply because it does not implement any authentication out of the box.

Reply:
Totally agree, I was being a slave to example in story #4380  Could you provide direction on following:

- Add option to disable all enhancers on rest server?  Alternatively option to override by providing array list of enhancers to apply?
- Would I follow similar pattern to [authentication strategies](https://github.com/strongloop/loopback4-example-shopping/tree/master/packages/shopping/src) say `spec-enhancers` folder alongside or would I seek to embed the enhancers within strategies.  If you have the time, perhaps a top level stepwise implementation path would be useful to follow.

Reply:
> Add option to disable all enhancers on rest server? Alternatively option to override by providing array list of enhancers to apply?

I am proposing to leave this part out of the initial implementation and wait until there is user demand for that.

> Would I follow similar pattern to [authentication strategies](https://github.com/strongloop/loopback4-example-shopping/tree/master/packages/shopping/src) say `spec-enhancers` folder alongside or would I seek to embed the enhancers within strategies.  If you have the time, perhaps a top level stepwise implementation path would be useful to follow.

IMO, it's not a concern of this pull request how the spec enhancer classes are registered.

I am expecting the `@loopback/authentication` extension to contribute a binding with `@bind(asSpecEnhancer)` decorator. As I understand the current implementation of `applyAllEnhancers`, it will automatically pick up such class.

See my ModelApiBooter spike in #3617 for an example. I really wish we had better documentation about the ExtensionPoint/Extension pattern :-/

Here the app decides to register a component (it's CrudRestComponent in the spike, it would be `@loopback/authentication` in your case):

https://github.com/strongloop/loopback-next/blob/d8fe4736fd07fc9da34ed345cbb77bdbd761b752/examples/todo/src/application.ts#L29

Here the component implements an Extension:

https://github.com/strongloop/loopback-next/blob/d8fe4736fd07fc9da34ed345cbb77bdbd761b752/packages/rest-crud/src/crud-rest-builder.plugin.ts#L30-L31

And here is the Extension implementation contributed by the Component to the Application context:

https://github.com/strongloop/loopback-next/blob/d8fe4736fd07fc9da34ed345cbb77bdbd761b752/packages/rest-crud/src/crud-rest.component.ts#L9-L11